### ISSUE TITLE
Consider warning on unintentional uses of the comma operator

### ISSUE THREAD

**Issue by nico** [llvm/llvm-project#18490]
|  |  |
| --- | --- |
| Bugzilla Link | [18116](https://llvm.org/bz18116) |
| Version | unspecified |
| OS | All |
| Attachments | [prototype patch](https://user-images.githubusercontent.com/3487/143748593-34bb683f-8a61-48c4-a63c-7d484ffe15fb.gz) |
| CC | @DougGregor |

## Extended Description 

In v8, some code called

   v8::Debug::SetLiveEditEnabled(false), env->GetIsolate();

when it meant to call

   v8::Debug::SetLiveEditEnabled(false, env->GetIsolate());

(SetLiveEditEnabled() has a default parameter for the 2nd argument)

A possible warning would be to check if the lhs of a comma operator is a call and the call is still valid when adding the rhs to the call arguments. The attached patch implements this idea – but it doesn't catch the same thing for overloads, and it didn't find any other bugs in chromium, so I'm not sure if it's worth it. I'm attaching the patch if someone else wants to do something with it.

**Comment by wheatman** [comment]
The warning `-Wcomma` was added in 3.9.
https://godbolt.org/z/nYGWfrWPo

code
```
int bar();
int foo(int a, int b = 5) {return a+b;}

int main() {
    foo(5),bar();
}
```
new warning
```
<source>:5:11: warning: possible misuse of comma operator here [-Wcomma]
    foo(5),bar();
          ^
<source>:5:5: note: cast expression to void to silence warning
    foo(5),bar();
    ^~~~~~
    static_cast<void>( )
1 warning generated.
Compiler returned: 0
```
